dex/dcr/script.go

```diff
@@ -560,10 +560,15 @@ func ExtractScriptAddrs(script []byte, chainParams *chaincfg.Params) (*DCRScript
 	pubkeys := make([]dcrutil.Address, 0)
 	pkHashes := make([]dcrutil.Address, 0)
 	// For P2SH and non-P2SH multi-sig, pull the addresses from the pubkey script.
-	_, addrs, numRequired, err := txscript.ExtractPkScriptAddrs(0, script, chainParams)
+	class, addrs, numRequired, err := txscript.ExtractPkScriptAddrs(0, script, chainParams)
 	if err != nil {
 		return nil, fmt.Errorf("ExtractScriptAddrs: %v", err)
 	}
+	// ExtractPkScriptAddrs will return a nil error and 0 addresses for a
+	// non-standard script, so we must check the class.
+	if class == txscript.NonStandardTy {
+		return nil, fmt.Errorf("non-standard script")
```

Comment:
This breaks `InputInfo`, which is obv a problem since `(*Backend).utxo` (for DCR) still is used to create an audited contract, a `asset.Contract`, in `Backend.Contract`.  But for the uses of `ExtractScriptAddrs` such as from `FundingCoin`>`utxo`>`InputInfo`, such callers need to recognize that the output is **not** usable to fund an order.

Reply:
Basically, I think `ExtractScriptAddrs` needs to return an error or the script class to the caller because 0 required addresses does not tell the whole story, and the needs of the caller are different (`Contract` vs. `FundingCoin`).

I added a check to dcr's `(*UTXO).Auth` that currently cannot be hit that checks for the zero addresses/required sigs case, which might be enough.  But this needs some thought since there may be other cases where zero and/or non-standard scripts are not acceptable.

Reply:
The `UTXO` type may need to diverge to two separate types too. 

I need to look again, but as a bonus, I think there is some optimization that can be done to the amount of information fetched and/or stored if we differentiate the two.

Reply:
Without changing too much, the `asset.Coin` interface (and `{dcr,btc}.UTXO`) could be given a `NonStandard() bool` method so that methods like `FundingCoin` can reject or accept the output they get from `(*Backend).utxo`.

Reply:
This needn't affect the `asset.Coin` interface. `FundingCoin` just needs to reject non-standard scripts.